### DECISION TREE CLASSIFIER

In [ ]:
# Best Decision Tree model: with max_depth adjusted AGAIN
pipe_forest = make_pipeline(StandardScaler(), DecisionTreeClassifier())
params = {'decisiontreeclassifier__max_depth' : [5,6,7,8],
          'decisiontreeclassifier__random_state' : [123]
          ,'decisiontreeclassifier__min_samples_split' : [2, 3, 4]
         }
grid_forest = GridSearchCV(pipe_forest, param_grid = params)

In [ ]:
grid_forest.fit(X_train,y_train)

GridSearchCV(estimator=Pipeline(steps=[('standardscaler', StandardScaler()),
                                       ('decisiontreeclassifier',
                                        DecisionTreeClassifier())]),
             param_grid={'decisiontreeclassifier__max_depth': [5, 6, 7, 8],
                         'decisiontreeclassifier__min_samples_split': [2, 3, 4],
                         'decisiontreeclassifier__random_state': [123]})

In [ ]:
grid_forest.score(X_test,y_test)### best DT model

0.6660015349194167

In [ ]:
pred = grid_forest.predict(X_test)

In [ ]:
f1_score(y_test,pred, average='micro')

0.6660015349194167

In [ ]:
grid_forest.best_estimator_

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('decisiontreeclassifier',
                 DecisionTreeClassifier(max_depth=7, min_samples_split=4,
                                        random_state=123))])

#### Best Decision Tree model hyperparameters:
    max_depth = 7
    min_samples_split=4
    Accuracy: 66.6%

#### Decision Tree models tested:
    With default parameters: 59.39%
    With max_depth = 5: 64.68%
    With max_depth = 7 & min_samples_split = 4: 66%
    With max_depth = 7 & min_samples_split = 7: 66.58%
        

### BAGGING CLASSIFIER

In [ ]:
pipe_bagged = make_pipeline(StandardScaler(), BaggingClassifier())
params = {
          'baggingclassifier__random_state' : [123]
          ,'baggingclassifier__max_features' : [9,10, 12, 14, 16, 20]
         }
grid_bagged = GridSearchCV(pipe_bagged, param_grid = params)

In [ ]:
grid_bagged.fit(X_train,y_train)

In [ ]:
############### best score ######################
grid_bagged.score(X_test,y_test)
0.6681504221028396

In [ ]:
pred = grid_bagged.predict(X_test)

f1_score(y_test,pred, average='micro')

In [ ]:
grid_bagged.best_estimator_

#### Best Bagging Classifier model hyperparameters:
    max_depth = 7
    min_samples_split=4
    Accuracy: 66.82%

#### Bagging Classifier models tested:¶
    With default parameters: 59.39%
    With max_features=10, max_samples=8, n_estimators=40: 56.13%
    With max_depth = 7 & min_samples_split = 4: 66%
    With max_depth = 7 & min_samples_split = 7: 66.58%